datalog_compiler/src/lib.rs

```diff
@@ -0,0 +1,545 @@
+use std::collections::{HashMap, HashSet};
+
+use hydroflow_lang::{
+    graph::flat_graph::FlatGraph,
+    parse::{
+        ArrowConnector, IndexInt, Indexing, NamedHfStatement, Operator, Pipeline, PipelineLink,
+    },
+};
+use proc_macro2::{Span, TokenStream};
+use quote::{quote, ToTokens};
+use syn::{parse_quote, punctuated::Punctuated, token::Paren, Token};
+
+mod grammar;
+
+use grammar::datalog::*;
+
+fn gen_datalog_program(literal: proc_macro2::Literal, root: TokenStream) -> syn::Stmt {
+    let str_node: syn::LitStr = parse_quote!(#literal);
+    let actual_str = str_node.value();
+    let program: Program = grammar::datalog::parse(&actual_str).unwrap();
+
+    let inputs = program
+        .rules
+        .iter()
+        .filter_map(|decl| match decl {
+            Declaration::Input(_, ident) => Some(ident.clone()),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    let outputs = program
+        .rules
+        .iter()
+        .filter_map(|decl| match decl {
+            Declaration::Output(_, ident) => Some(ident.clone()),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    let rules = program
+        .rules
+        .iter()
+        .filter_map(|decl| match decl {
+            Declaration::Rule(rule) => Some(rule.clone()),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
```

Comment:
I'd prefer a single for loop here, or use https://docs.rs/separable/latest/separable/ if you wanna be fancy

Reply:
Also is the clone needed? Shared references should be ok for the lifetime of the fn?